docs/notebooks/code_overview.pct.py

```diff
@@ -0,0 +1,238 @@
+# %% [markdown]
+# # An overview of Trieste types
+
+# %% [markdown]
+# Trieste is dedicated to Bayesian optimization, the process of finding the *optimal values of an expensive, black-box objective function by employing probabilistic models over observations*. This notebook explains how the different parts of this process are represented by different types in the code, and how these types can be extended.
+
+# %% [markdown]
+# ## Key types
+#
+# The following types represent the key concepts in Trieste. For a full listing of all the types in Trieste, see the API reference.
+
+# %% [markdown]
+# ### `Observer`
+#
+# The `Observer` type definition represents the black box objective function. Observers are functions that accept query points and return observations. Observations are either a single objective value that we wish to optimize, or a dictionary of mutiple tagged values that must be combined somehow, for example an objective and an inequality constraint. Objective values can be either single or multi-dimensional (see [multi-objective optimization](multi_objective_ehvi.ipynb)).
```

Comment:
Observers return a dataset rather than observations, which is slightly unusual so we should state it.

Not sure how to write it, but it might be worth noting that this is very flexible in the sense that the returned dataset can be anything (not necessarily [query points, observations]).